packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1164 @@
+      final Offset focalPointScene = _transformationController.toScene(
+        event.localPosition,
+      );
+      _transformationController.value = _matrixScale(
+        _transformationController.value,
+        scaleChange,
+      );
+
+      // After scaling, translate such that the event's position is at the
+      // same scene point before and after the scale.
+      final Offset focalPointSceneScaled = _transformationController.toScene(
+        event.localPosition,
+      );
+      _transformationController.value = _matrixTranslate(
+        _transformationController.value,
+        focalPointSceneScaled - focalPointScene,
+      );
+    }
+  }
+
+  // Handle inertia drag animation.
+  void _onAnimate() {
+    if (!_controller.isAnimating) {
+      _animation?.removeListener(_onAnimate);
+      _animation = null;
+      _controller.reset();
+      return;
+    }
+    // Translate such that the resulting translation is _animation.value.
+    final Vector3 translationVector = _transformationController.value.getTranslation();
+    final Offset translation = Offset(translationVector.x, translationVector.y);
+    final Offset translationScene = _transformationController.toScene(
+      translation,
+    );
+    final Offset animationScene = _transformationController.toScene(
+      _animation.value,
+    );
+    final Offset translationChangeScene = animationScene - translationScene;
+    _transformationController.value = _matrixTranslate(
+      _transformationController.value,
+      translationChangeScene,
+    );
+  }
+
+  void _onTransformationControllerChange() {
+    // A change to the TransformationController's value is a change to the
+    // state.
+    setState(() {});
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    _transformationController = widget.transformationController
+        ?? TransformationController();
+    _transformationController.addListener(_onTransformationControllerChange);
+    _controller = AnimationController(
+      vsync: this,
+    );
+  }
+
+  @override
+  void didUpdateWidget(InteractiveViewer oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.child != oldWidget.child || widget.boundaryMargin != oldWidget.boundaryMargin) {
+      _boundaryRectCached = null;
+    }
+
+    // Handle all cases of needing to dispose and initialize
+    // transformationControllers.
+    if (oldWidget.transformationController == null) {
+      if (widget.transformationController != null) {
+        _transformationController.removeListener(_onTransformationControllerChange);
+        _transformationController.dispose();
+        _transformationController = widget.transformationController;
+        _transformationController.addListener(_onTransformationControllerChange);
+      }
+    } else {
+      if (widget.transformationController == null) {
+        _transformationController.removeListener(_onTransformationControllerChange);
+        _transformationController.dispose();
```

Comment:
In this case, the controller is owned by oldWidget, right? We shouldn't dispose it here.

Reply:
Yes, good catch.